### PR TITLE
feat(rpc): Implement `getblockchaininfo` RPC method

### LINKED ISSUE #3143: getblockchaininfo JSON-RPC method
## Motivation

`lightwalletd` uses the `getblockchaininfo` JSON-RPC method.

### Dependencies

- [x] #3161

### Required Fields

The method is documented here: https://zcash.github.io/rpc/getblockchaininfo.html

But only the following fields are used by `lightwalletd`: https://github.com/zcash/lightwalletd/blob/master/common/common.go#L70

Field list:
```
{
  "chain": "xxxx",        (string) current network name as defined in BIP70 (main, test, regtest)
  "blocks": xxxxxx,         (numeric) the current number of blocks processed in the server
  "bestblockhash": "...", (string) the hash of the currently best block
  "estimatedheight": xxxx,  (numeric) if syncing, the estimated height of the chain, else the current best height
  "upgrades": {                (object) status of network upgrades
     "xxxx" : {                (string) branch ID of the upgrade, in lowercase hexadecimal
        "name": "xxxx",        (string) name of upgrade - unused by `lightwalletd`, but useful for debugging
        "activationheight": xxxxxx,  (numeric) block height of activation
        "status": "xxxx",      (string) status of upgrade: disabled/pending/active
     }, ...
  },
  "consensus": {               (object) branch IDs of the current and upcoming consensus rules
     "chaintip": "xxxxxxxx",   (string) branch ID used to validate the current chain tip, in lowercase hexadecimal
     "nextblock": "xxxxxxxx"   (string) branch ID that the next block will be validated under, in lowercase hexadecimal
  }
}
```

The `zcashd` code that generates the network upgrade JSON is here:
https://github.com/zcash/zcash/blob/6cd5b8792bbe0a3cc1064e1884914415f0c6e7d7/src/rpc/blockchain.cpp#L970

The `lightwalletd` code that parses the network upgrade JSON is here:
https://github.com/adityapk00/lightwalletd/blob/c1bab818a683e4de69cd952317000f9bb2932274/common/common.go#L195

### Tasks

- Implement the RPC method 
- Add RPC acceptance tests to CI
- Test that the RPC method works with `lightwalletd` 

## API Reference

We plan to use [jsonrpc_core](https://docs.rs/jsonrpc-core/18.0.0/jsonrpc_core/) with:
* [jsonrpc_http_server](https://github.com/paritytech/jsonrpc/tree/master/http)
* [jsonrpc-derive](https://github.com/paritytech/jsonrpc/tree/master/derive)

### Example Code

Here are examples of:
* [declaring RPC methods as Rust functions](https://github.com/paritytech/jsonrpc#basic-usage-with-derive)

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Close https://github.com/ZcashFoundation/zebra/issues/3143

## Solution

Implement `getblockchaininfo`

## Review

Tests are missing. I am thinking on them now and i will remove from draft then. Anyone who want to take a look before then is welcome. 

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-chain/src/parameters/network_upgrade.rs
Comment (teor2345) [inline zebra-chain/src/parameters/network_upgrade.rs]: Optional simplification

```suggestion
        self.0.to_be_bytes()
```

File: zebra-chain/src/parameters/network_upgrade.rs
Comment (teor2345) [inline zebra-chain/src/parameters/network_upgrade.rs]: Question: why are the bytes reversed in ToHex, but not in FromHex?

**Review by teor2345 — COMMENTED** [review]
I think @conradoplg is fixing this in another PR?

See the chat in Discord.

File: zebra-chain/src/parameters/network_upgrade.rs
Comment (oxarbitrage) [reply zebra-chain/src/parameters/network_upgrade.rs]: https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/hash.rs#L32-L35

File: zebra-chain/src/parameters/network_upgrade.rs
Comment (oxarbitrage) [reply zebra-chain/src/parameters/network_upgrade.rs]: https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/hash.rs#L53-L61

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/hash.rs#L73-L81

Comment (oxarbitrage) [reply zebra-chain/src/parameters/network_upgrade.rs]: We should probably change above as well. Ill wait for @conradoplg before making changes here.

Comment (conradoplg) [reply zebra-chain/src/parameters/network_upgrade.rs]: Created a separate PR for it if you need it here: https://github.com/ZcashFoundation/zebra/pull/3893

**Comment by oxarbitrage** [comment]
Might be blocked on https://github.com/ZcashFoundation/zebra/pull/3899 (which also depends on others) because we want to know why the lightwalletd is failing (https://github.com/ZcashFoundation/zebra/runs/5590293910?check_suite_focus=true#step:4:253) and potentially make changes if needed. This should be easier to do with better diagnostics that are in those series.

In an unrelated issue now it seems all tests that are in the cloud are failing with this error: 

```
Error response from daemon: manifest for us-docker.pkg.dev/zealous-zebra/zebra/zebrad-test:sha-c015a374 not found: manifest unknown: Failed to fetch "sha-c015a374"
Error: Process completed with exit code 1.
```
https://github.com/ZcashFoundation/zebra/runs/5618372108?check_suite_focus=true

Happening today in other pull requests, the same error but different commit hash: https://github.com/ZcashFoundation/zebra/runs/5618335743?check_suite_focus=true

I have no idea what is this or how to bypass.  @gustavovalverde can you take a look?

**Comment by teor2345** [comment]
> Happening today in other pull requests, the same error but different commit hash: https://github.com/ZcashFoundation/zebra/runs/5618335743?check_suite_focus=true
> 
> I have no idea what is this or how to bypass. @gustavovalverde can you take a look?

I'm wondering if PR #3885 caused this issue, it's the last big CI change that merged, and the last PR that passed GitHub CI.

(PR #3907 probably passed GitHub CI without PR #3885, then passed Mergify CI with PR #3885.)

**Comment by gustavovalverde** [comment]
@teor2345 this PR does not have the latest changes from https://github.com/ZcashFoundation/zebra/pull/3885 but the error might still happen. Can you rebase this branch with `main` @oxarbitrage ?

**Comment by teor2345** [comment]
> @teor2345 this PR does not have the latest changes from #3885 but the error might still happen. Can you rebase this branch with `main` @oxarbitrage ?

@gustavovalverde these bugs still happen for PRs based on main, for both the initial GitHub jobs and the Mergify queues. For example:
- https://github.com/ZcashFoundation/zebra/pull/3874
- https://github.com/ZcashFoundation/zebra/pull/3913

Did you want to revert PR #3885?

**Comment by teor2345** [comment]
> Might be blocked on #3899 (which also depends on others) because we want to know why the lightwalletd is failing (https://github.com/ZcashFoundation/zebra/runs/5590293910?check_suite_focus=true#step:4:253) and potentially make changes if needed. This should be easier to do with better diagnostics that are in those series.

I updated the integration tests in PR #3914. There's no need to block on better diagnostics.

There is a behaviour change in these lightwalletd logs:
> {"app":"lightwalletd","level":"info","msg":"Got sapling height 419200 block height 2000 chain main branchID 00000000","time":"2022-03-17T18:36:56Z"}
> {"app":"lightwalletd","error":"error requesting block: 0: Block not found","height":419200,"level":"warning","msg":"error zcashd getblock rpc","time":"2022-03-17T18:36:56Z"}

https://github.com/ZcashFoundation/zebra/runs/5590293910?check_suite_focus=true#step:4:230

Before we implemented getblockchaininfo, lightwalletd was getting Sapling height 0 (due to Go json missing fields) and fetching block 0 onwards.

Now it is correctly loading blocks from Sapling activation onwards.

**Comment by gustavovalverde** [comment]
Found the issue in one of the actions as it's version number was not fixed and a bug was recently introduced https://github.com/rlespinasse/github-slug-action/issues/92 

I'll be creating a PR to pin the action version. Re-running the Action might be a workaround in the meanwhile.

**Review by teor2345 — COMMENTED** [review]
I opened these PRs:

Required for tests to work:
* #3914

Optional suggestions:
* #3915

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs:308]: Optional: We're using GetBestBlockHash by itself and as part of GetBlockChainInfo. This is a bit confusing, because it's not a get_best_block_hash response here.

Can we rename it based on what it is, rather than how it is used?

```suggestion
            best_block_hash: BlockHashHex(tip_hash),
```

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Concurrency Question: What happens if the tip block changes between the height and hash function calls?

(The estimated height doesn't matter as much, because it's just an estimate.)

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Incorrect Value: The definition of this field is:
> if syncing, the estimated height of the chain, else the current best height

Can we follow this definition, and check if Zebra is still syncing?
I think lightwalletd might exit with a fatal error if we don't.

If lightwalletd expects more blocks, but it can't get any for a few minutes, it exits. The estimate can be a long way ahead of the actual tip, because it's based on the block time, which can vary by up to 2 hours from the local time.

So if the estimate is a long way ahead, and the network goes down for a few minutes, lightwalletd might exit.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs:536]: Missing Docs: The field definitions aren't in this file, so that makes it harder to review changes.

Do you think we should document each field in this struct, and every struct and enum it uses?

Copying the field definitions from the ticket would really help me do reviews.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs:527]: We also missed documenting the fields in the `GetInfo` struct.

Do you think we should document them too?

**Review by teor2345 — COMMENTED** [review]
I'd like to fix the estimated tip value, and make sure the tip hash and height are consistent. 

Everything else is optional.

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: https://github.com/ZcashFoundation/zebra/pull/3891/commits/aebae9cf52268360a958e7f5caa255df285452ee

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: Yea, you are right, i missed that. 

I was wondering how to do this. Can we just compare? Something as:

```
if tip_height > estimated_height {
   estimated_height = tip_height;
}
else {
   estimated_height = estimated_height;
}
```

Or do we need to go deeper and create an additional method in the chaintip ?

By the way, looking at `estimate_network_chain_tip_height()` definition i don't know why it returns an `Option` if it always return `Some()`. https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/chain_tip.rs#L46-L57
So another option may be to return `None` from there if we are already in the tip (i am not sure how hard will be to implement that)

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs]: Thanks!

Maybe we should just return the whole `ChainTipBlock` instead, but we can do that cleanup later.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs]: > looking at `estimate_network_chain_tip_height()` definition i don't know why it returns an `Option` if it always return `Some()`

The `?` operator on this line returns `None` if `best_tip_height_and_block_time()` returns `None`.
https://github.com/ZcashFoundation/zebra/blob/5d7f9861830ce4f5de9d5d68df15854a56610f55/zebra-chain/src/chain_tip.rs#L51

Comment (teor2345) [reply zebra-rpc/src/methods.rs]: > Or do we need to go deeper and create an additional method in the chaintip ?

### Priority

I can't see where this field is used in lightwalletd or ZecWallet Lite. So we might just be able to open a ticket and fix it later.

I've just sent an email to Aditya to double-check.

### Specification

I checked what `zcashd` does here:
1. If it has stopped syncing, return the current height: https://github.com/zcash/zcash/blob/6cd5b8792bbe0a3cc1064e1884914415f0c6e7d7/src/rpc/blockchain.cpp#L1068-L1071
2. If the block time is ahead of the current time, return the current height: https://github.com/zcash/zcash/blob/f7006e9a33ca37ae88520e434ee5c2b2498c1cc6/src/metrics.cpp#L141-L143
3. Otherwise, estimate the tip height: https://github.com/zcash/zcash/blob/f7006e9a33ca37ae88520e434ee5c2b2498c1cc6/src/metrics.cpp#L145-L160

Here's the `SyncStatus` method that returns if we're close to the tip:
https://github.com/ZcashFoundation/zebra/blob/5d7f9861830ce4f5de9d5d68df15854a56610f55/zebrad/src/components/sync/status.rs#L46-L47

### Zebra Changes

Here's one way to do this fix:
1. Move `SyncStatus` to `zebra_node_services`
2. Add `SyncStatus` to `RpcImpl`, and use it to check if we're close to the tip
3. Fix `NetworkChainTipHeightEstimator` to return the current height if the block time is ahead of the current time

Here's the `NetworkChainTipHeightEstimator` code. Removing negative times should simplify it a lot. The bug fix will probably make some tests fail, Janito can help you fix them.

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/chain_tip/network_chain_tip_height_estimator.rs#L111-L120

Comment (teor2345) [reply zebra-rpc/src/methods.rs]: Aditya said a rough estimate was ok.

So let's go with a simple fix for now. I will open a ticket for the rest.

What do you think of these changes?
1. If the block time is ahead of the current time, return the current height
2. If the estimate is below the current height, return the current height

Comment (teor2345) [reply zebra-rpc/src/methods.rs]: I put the rest of the changes in #3954

Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: Added the simple fix in https://github.com/ZcashFoundation/zebra/pull/3891/commits/56f6ba37ad350ca35b3208139ef0b3e9a94ff59a

File: zebra-chain/src/parameters/tests.rs
Comment (jvff) [inline zebra-chain/src/parameters/tests.rs:250]: Minor: I'd also add another check, just to be sure.

```suggestion
            prop_assert_eq!(branch, new_branch);
            prop_assert_eq!(hex_branch, new_branch.to_string());
```

To apply this you'll also need to remove the `&` from `&nu.branch_id()`.

File: zebra-rpc/src/methods.rs
Comment (jvff) [inline zebra-rpc/src/methods.rs]: Typo:

```suggestion
                // zcashd's RPC seems to ignore Disabled network upgrades, so Zebra does too.
```

File: zebra-rpc/src/methods/tests/prop.rs
Comment (jvff) [inline zebra-rpc/src/methods/tests/prop.rs]: Nit: this seems like two tests in one. I'd suggest to split it in two:

```suggestion
    /// Test the `get_blockchain_info` response when Zebra's state is empty.
    #[test]
    fn get_blockchain_info_response_without_a_chain_tip(network in any::<Network>()) {
        let runtime = zebra_test::init_async();
        let _guard = runtime.enter();

        let mut mempool = MockService::build().for_prop_tests();
        let mut state: MockService<_, _, _, BoxError> = MockService::build().for_prop_tests();

        // look for an error with a `NoChainTip`
        let rpc = RpcImpl::new(
            "RPC test",
            Buffer::new(mempool.clone(), 1),
            Buffer::new(state.clone(), 1),
            NoChainTip,
            network,
        );
        let response = rpc.get_blockchain_info();
        prop_assert_eq!(&response.err().unwrap().message, "No Chain tip available yet");

        runtime.block_on(async move {
            mempool.expect_no_requests().await?;
            state.expect_no_requests().await?;

            Ok::<_, TestCaseError>(())
        })?;
    }
    
    /// Test the `get_blockchain_info` response using an arbitrary block as the `ChainTip`.
    #[test]
    fn get_blockchain_info_response_with_an_arbitrary_chain_tip(
        network in any::<Network>(),
        block in any::<Block>(),
    ) {
        let runtime = zebra_test::init_async();
        let _guard = runtime.enter();

        let mut mempool = MockService::build().for_prop_tests();
        let mut state: MockService<_, _, _, BoxError> = MockService::build().for_prop_tests();

        // get block data
```

**Review by jvff — DISMISSED** [review]
Looks great! I've only added a few optional suggestions.

File: zebra-chain/src/parameters/tests.rs
Comment (oxarbitrage) [reply zebra-chain/src/parameters/tests.rs:250]: https://github.com/ZcashFoundation/zebra/pull/3891/commits/c601c8ab3ed0c4b84a64b26cc2aac6507094c962

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: https://github.com/ZcashFoundation/zebra/pull/3891/commits/d14e104d0ff3b1af6f95aa31a61f64e7e4380024

File: zebra-rpc/src/methods/tests/prop.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods/tests/prop.rs]: https://github.com/ZcashFoundation/zebra/pull/3891/commits/780fca792cd718a76478438a595b2554bd0dc4d7

**Comment by oxarbitrage** [comment]
@teor2345 can you check the lightwalletd test failure with this pull request?

https://github.com/ZcashFoundation/zebra/runs/5678177141?check_suite_focus=true#step:3:226
```
Logged a failure message:
{"app":"lightwalletd","error":"0: No Chain tip available yet","level":"warning","msg":"error with getblockchaininfo rpc, retrying...","retry":1,"time":"2022-03-24T14:48:27Z"}

Matching failure regex: [
    "error with",
]

All Failure regexes: [
    "panic:",
    "unable to issue RPC call",
    "Invalid params",
    "Method not found",
    "Lightwalletd died with a Fatal error",
    "json: cannot unmarshal",
    "into Go value of type",
    "block requested is newer than latest block",
    "Cache add failed",
    "error decoding",
    "error marshaling",
    "error parsing JSON",
    "error reading JSON response",
    "error with",
    "received overlong message",
    "received unexpected height block",
    "Reorg exceeded max",
    "unable to issue RPC call",
    "Aborted",
    "Abort trap",
]

Location: zebra-test/src/command.rs:[235](https://github.com/ZcashFoundation/zebra/runs/5678177141?check_suite_focus=true#step:3:235)
```

I am unsure on how to fix it. Thanks!

**Review by jvff — DISMISSED** [review]

**Comment by teor2345** [comment]
> {"app":"lightwalletd","error":"0: No Chain tip available yet","level":"warning","msg":"error with getblockchaininfo rpc, retrying...","retry":1

I just checked what `zcashd` does here, and it also returns an error just after startup. I'll try to adjust the regex to ignore the error for this RPC, but still catch most other errors.

**Comment by teor2345** [comment]
> > {"app":"lightwalletd","error":"0: No Chain tip available yet","level":"warning","msg":"error with getblockchaininfo rpc, retrying...","retry":1
> 
> I just checked what `zcashd` does here, and it also returns an error just after startup. I'll try to adjust the regex to ignore the error for this RPC, but still catch most other errors.

@oxarbitrage I opened PR #3961, let me know what you think?

**Comment by oxarbitrage** [comment]
I merged https://github.com/ZcashFoundation/zebra/pull/3961, please merge this one if the CI pass. Thank you.

**Comment by teor2345** [comment]
@mergifyio update

**Review by teor2345 — APPROVED** [review]
Looks good, and there's a bunch of changes here that other people will want.

**Pull request merged**
